sources/packages/backend/libs/services/src/msfaa-number/msfaa-number-shared.service.ts

```diff
@@ -0,0 +1,269 @@
+      );
+    if (!pendingDisbursement) {
+      throw new CustomNamedError(
+        "Not possible to reissue an MSFAA when there is no pending disbursements for the application.",
+        APPLICATION_INVALID_DATA_TO_CREATE_MSFAA_ERROR,
+      );
+    }
+    if (!pendingDisbursement.msfaaNumber.cancelledDate) {
+      throw new CustomNamedError(
+        "Not possible to reissue an MSFAA when the current associated MSFAA is not cancelled.",
+        APPLICATION_INVALID_DATA_TO_CREATE_MSFAA_ERROR,
+      );
+    }
+    return this.internalCreateMSFAANumber(
+      application.student.id,
+      application.id,
+      application.currentAssessment.offering.offeringIntensity,
+      auditUserId,
+    );
+  }
+
+  /**
+   * Creates a new MSFAA record with a new number for the specified student.
+   * @param studentId student to have a new MSFAA record created.
+   * @param referenceApplicationId reference application id.
+   * @param offeringIntensity offering intensity.
+   * @param auditUserId user that should be considered the one that is causing the changes.
+   * individually based on, for instance, the Part time/Full time.
+   * @returns created MSFAA record.
+   */
+  private async internalCreateMSFAANumber(
+    studentId: number,
+    referenceApplicationId: number,
+    offeringIntensity: OfferingIntensity,
+    auditUserId: number,
+  ): Promise<MSFAANumber> {
+    return this.dataSource.transaction(async (entityManager) => {
+      const auditUser = await this.systemUsersService.systemUser();
+      const now = new Date();
+      const nowISODate = getISODateOnlyString(now);
+      // Cancel any pending MSFAA record that was never reported as signed or cancelled.
+      // This will ensure that the newly created MSFAA will be the only one valid.
+      await entityManager.getRepository(MSFAANumber).update(
+        {
+          student: { id: studentId },
+          offeringIntensity,
+          dateSigned: IsNull(),
+          cancelledDate: IsNull(),
+        },
+        {
+          cancelledDate: nowISODate,
+          modifier: auditUser,
+          updatedAt: now,
+        },
+      );
+      // Create the new MSFAA record for the student.
+      const newMSFAANumber = new MSFAANumber();
+      newMSFAANumber.msfaaNumber = await this.consumeNextSequence(
+        offeringIntensity,
+      );
+      newMSFAANumber.student = { id: studentId } as Student;
+      newMSFAANumber.referenceApplication = {
+        id: referenceApplicationId,
+      } as Application;
+      newMSFAANumber.offeringIntensity = offeringIntensity;
+      newMSFAANumber.creator = { id: auditUserId } as User;
+      newMSFAANumber.createdAt = now;
+      await entityManager.getRepository(MSFAANumber).save(newMSFAANumber);
+      // Associate pending disbursements with the new MSFAA.
+      const disbursementScheduleRepo =
+        entityManager.getRepository(DisbursementSchedule);
+      const schedulesToUpdate = await disbursementScheduleRepo.find({
```

Comment:
We can directly use update like this right?

```ts

await disbursementScheduleRepo.update(
        {
          disbursementScheduleStatus: DisbursementScheduleStatus.Pending,
          studentAssessment: {
            application: { student: { id: studentId } },
            offering: { offeringIntensity: offeringIntensity },
          },
        },
        {
          msfaaNumber: newMSFAANumber,
          modifier: auditUser,
          updatedAt: now,
        },
      );
```

Reply:
This was my first attempt. Typeorm does not allow it apparently and, unless I am missing something, this is probably the reason why we do not have any other similar update like this.
Please notice that the main difference is that the "where" clause for this would include relations.